.github/workflows/ci.yml

```
@@ -4,27 +4,77 @@ on:
   pull_request:
   merge_group:
 
-permissions:
-  contents: read
-  id-token: write
-
 jobs:
-  ci:
-    name: CI
+  build:
+    name: Build and test
     runs-on: ubuntu-latest
+    permissions:
+      contents: read
     steps:
       - name: Checkout the source code
         uses: actions/checkout@v4
 
       - name: Check formatting
         run: cargo fmt --check
 
+      - name: Build the tool
+        run: cargo build
+
+      - name: Try a dry run of the changes
+        run: cargo run -- --skip-upload
+        if: github.event_name == 'pull_request'
+
+      - name: Upload tool to artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: tool
+          path: target/debug/ci-mirrors
+          if-no-files-found: error
+
+      - name: Upload files.toml to artifacts
+        uses: actions/upload-artifact@v4
+        with:
+          name: files-manifest
+          path: files.toml
+          if-no-files-found: error
+
+  apply:
+    name: Apply to production
+    runs-on: ubuntu-latest
+    needs: [build]
+    if: github.event_name == 'merge_queue'
```

```suggestion
    if: github.event_name == 'merge_group'
```
I always get this wrong :laughing: